### PR TITLE
Implement Win64 eh_personality natively.

### PR DESCRIPTION
**Pull request by vadimcn**
After this change, the only remaining symbol we are pulling from libgcc on Win64 is `__chkstk_ms` - the stack probing routine.

### CODE REVIEW

**Comment by vadimcn** [comment]
I am not quite happy about the way `rust_eh_unwind_resume` get linked.  It should have probably been a lang item, but I couldn't quite figure out how to plumb it through to `LLVMRustHookUnwindResume`...

r? @alexcrichton

File: src/doc/trpl/no-stdlib.md [hunk 1]
```diff
@@ -71,18 +73,18 @@ The compiler currently makes a few assumptions about symbols which are available
...
 maintained then a thread always has an "infinite stack" and this function
 shouldn't get triggered.
 
-The second of these three functions, `eh_personality`, is used by the
-failure mechanisms of the compiler. This is often mapped to GCC's
-personality function (see the
-[libstd implementation](../std/rt/unwind/index.html) for more
+The second and third of these functions, `eh_personality` and `rust_eh_unwind_resume`,
+are used by the failure mechanisms of the compiler. These are often mapped to GCC's
+personality function and libgcc's `_Unwind_Resume` respectively
```
Comment (nagisa) [inline src/doc/trpl/no-stdlib.md]: Wording of this sentence somehow seems to suggest GCC personality function and `_Unwind_Resume` come from two different locations, while in practice they both come from `libgcc`.

File: src/librustc_trans/back/write.rs [hunk 2]
```diff
@@ -512,6 +517,13 @@ unsafe fn optimize_and_codegen(cgcx: &CodegenContext,
         None => {},
     }
 
+    // Inject a local-scoped `_Unwind_Resume` function, which redirects to `rust_eh_unwind_resume`.
+    // This must be done after optimizations, because until DwarfEHPrepare pass has been run,
+    // `_Unwind_Resume` is not referenced by any live code and would have been eliminated by opt.
```
Comment (nagisa) [inline src/librustc_trans/back/write.rs]: Wouldn’t [OptimizeNone](http://llvm.org/docs/doxygen/html/Attributes_8h_source.html#l00098) prevent it from being eliminated?

Comment (alexcrichton) [inline src/librustc_trans/back/write.rs]: This may also be able to prevent elimination through the [`llvm.used` global](http://llvm.org/docs/LangRef.html#the-llvm-used-global-variable)

File: src/doc/trpl/no-stdlib.md [hunk 3]
```diff
@@ -38,6 +38,7 @@ fn start(_argc: isize, _argv: *const *const u8) -> isize {
 // provided by libstd.
 #[lang = "stack_exhausted"] extern fn stack_exhausted() {}
 #[lang = "eh_personality"] extern fn eh_personality() {}
+#[no_mangle] pub extern fn rust_eh_unwind_resume() {}
```
Comment (alexcrichton) [inline src/doc/trpl/no-stdlib.md]: It looks like this only needs to be defined on 64-bit windows targets, so can this be omitted? I suspect 99% of `#![no_std]` users aren't working on Windows.

File: src/librustc_back/target/x86_64_pc_windows_gnu.rs [hunk 4]
```diff
@@ -16,6 +16,7 @@ pub fn target() -> Target {
     // On Win64 unwinding is handled by the OS, so we can link libgcc statically.
     base.pre_link_args.push("-static-libgcc".to_string());
     base.pre_link_args.push("-m64".to_string());
+    base.custom_unwind_resume = true;
```
Comment (alexcrichton) [inline src/librustc_back/target/x86_64_pc_windows_gnu.rs]: Should this also set the hook for x86_64-pc-windows-msvc?

File: src/libstd/rt/dwarf/eh.rs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+#[derive(Copy, Clone)]
+pub struct EHContext {
+    pub ip: usize,         // Current instruction pointer
+    pub func_start: usize, // Address of the current function
+    pub text_start: usize, // Address of the code section
+    pub data_start: usize, // Address of the data section
+}
+
+#[inline(never)]
+pub unsafe fn find_landing_pad(lsda: *const u8, context: EHContext) -> Option<usize> {
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: A few notes on this function:

* Is there public documentation available for this protocol? I'd love to reference this against a known implementation or description to verify it, but also so we can understand what this is doing months from now.
* Can you document why this is `inline(never)`?

File: src/libstd/rt/dwarf/mod.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! General DWARF format parsing utilities
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/mod.rs]: Similarly as above, it'd be great to have a lot more documentation on this module. For example the `read_uleb128` function would be nice to have a hyperlink to the definition of the format, and some overall docs about what this module's doing would be nice.

File: src/libstd/rt/unwind/gcc.rs [hunk 7]
```diff
@@ -129,7 +120,10 @@ pub mod eabi {
             uw::_URC_HANDLER_FOUND // catch!
         }
         else { // cleanup phase
-            uw::_URC_INSTALL_CONTEXT
+            unsafe {
+                __gcc_personality_v0(version, actions, exception_class, ue_header,
+                                     context)
+            }
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/gcc.rs]: Just to clarify, was this a drive-by fix?

File: src/libstd/rt/unwind/gcc.rs [hunk 8]
```diff
@@ -99,8 +92,6 @@ pub mod eabi {
     }
 
     #[lang="eh_personality"]
-    #[no_mangle] // referenced from rust_try.ll
-    #[allow(private_no_mangle_fns)]
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/gcc.rs]: Does this pass all the tests? I vaguely remember seeing the `rust_eh_personality` symbol in a list of reachable symbols which was necessary for LTO working, but I may be remembering incorrectly.

File: src/libstd/rt/mod.rs [hunk 9]
```diff
@@ -47,6 +47,9 @@ pub mod args;
 mod at_exit_imp;
 mod libunwind;
 
+#[cfg(all(windows, target_arch="x86_64", target_env="gnu"))]
```
Comment (alexcrichton) [inline src/libstd/rt/mod.rs]: It may be worth having this just be compiled for all platforms and it'll just end up being optimized away on most. That way we can catch breakage to it on all platforms instead of just 64-bit windows.

File: src/libstd/rt/unwind/gcc.rs (hunk 7 above)
Comment (alexcrichton) [inline src/libstd/rt/unwind/gcc.rs]: Oh hm ok I think I see what's going on here, it's because this is legit running in something other than the search phase, so we probably need to do something more than just returning a constant.

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+// Since this cannot be done with libdwarf-style landing pads, rust_eh_personality instead
+// catches RUST_PANICs, runs the landing pad, then reraises the exception.
+//
+// Note that it makes certain assumptions about the exception:
+// 1. That RUST_PANIC is non-continuable, so no lower stack frame may choose to resume execution.
+// 2. That the first parameter of the exception is a pointer to an extra data area (PanicData).
+// Since these assumptions do not generally hold true for foreign exceptions (system faults,
+// C++ exceptions, etc), we make no attempt to invoke our landing pads (and, thus, destructors!)
+// for anything other than RUST_PANICs. This is considered acceptable, because the behavior of
+// throwing exceptions through a C ABI boundary is undefined.
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Could you format this to 80-chars wide? (what the rest of the runtime is formatted to)

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+) -> EXCEPTION_DISPOSITION
+{
+    let er = &*exceptionRecord;
+    let dc = &*dispatcherContext;
+    rtdebug!("rust_eh_personality_catch: code={:X}, flags={:X}, frame={:X}, ip={:X}",
+        er.ExceptionCode, er.ExceptionFlags, establisherFrame as usize, dc.ControlPc as usize);
+
+    if er.ExceptionFlags & EXCEPTION_UNWIND == 0 { // we are in the dispatch phase
+        if er.ExceptionCode == RUST_PANIC {
+            if let Some(lpad) = find_landing_pad(dc) {
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: This may be a bit clearer as:

```
let lpad = find_landing_pad(dc).unwrap_or_else(|| {
    rtabort!(...)
});
```

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+            }
+            rtabort!("could not locate the landing pad");
+        }
+    }
+    ExceptionContinueSearch
+}
+
+#[lang="eh_personality"]
+#[cfg(not(test))]
+unsafe extern "C" fn rust_eh_personality(
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Is this just the same as `rust_eh_personality_catch`? These two functions look the same except for the expectation of the landing pad existing...

Overall I'd find it at least helpful to have a nice big doc comment at the top of the module explaining what's going on here. I have a little understanding of what mechanisms are in play, but I know why these are structured the way they are. For example it's not clear to my why the catch personality is catching but this one isn't?

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+                            contextRecord,
+                            dc.HistoryTable);
+                rtabort!("could not unwind");
+            }
+        }
+    }
+    ExceptionContinueSearch
+}
+
+#[no_mangle] // referenced from landing pads
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Can you expand this to a comment explaining what this function is doing? It'd be nice to mention that this is called from all landing pads and is injected by LLVM, and it's also why it's `pub` + `#[no_mangle]`.

Also, can you remove the `"C"` part from the header?

**Comment by alexcrichton** [comment]
Nice work @vadimcn!

File: src/doc/trpl/no-stdlib.md (hunk 3 above)
Comment (vadimcn) [inline src/doc/trpl/no-stdlib.md]: Without it doctests fail on Windows.  Also, if we ever drop libgcc on other platforms, we'll need it.

File: src/librustc_back/target/x86_64_pc_windows_gnu.rs (hunk 4 above)
Comment (vadimcn) [inline src/librustc_back/target/x86_64_pc_windows_gnu.rs]: I was thinking of a separate PR.

File: src/libstd/rt/dwarf/eh.rs (hunk 5 above)
Comment (vadimcn) [inline src/libstd/rt/dwarf/eh.rs]: > Is there public documentation available for this protocol? 

Not really.  Gcc LSDA format is defined by, well, gcc implementation.  Aside from that, there are a few blog posts...  I'll link them here.

File: src/libstd/rt/dwarf/mod.rs (hunk 6 above)
Comment (vadimcn) [inline src/libstd/rt/dwarf/mod.rs]: Now these are defined in the DWARF standard.  I'll add a section reference here.

File: src/libstd/rt/dwarf/eh.rs (hunk 5 above)
Comment (vadimcn) [inline src/libstd/rt/dwarf/eh.rs]: inline(never) isn't needed, just forgot to delete it

File: src/libstd/rt/unwind/gcc.rs (hunk 8 above)
Comment (vadimcn) [inline src/libstd/rt/unwind/gcc.rs]: It did, on Windows.  Might be worth pushing this into try?

File: src/libstd/rt/unwind/gcc.rs (hunk 7 above)
Comment (vadimcn) [inline src/libstd/rt/unwind/gcc.rs]: This is basically a roll-back of the rust_try_inner magic, which was created because of the way libgcc's C personality routine works on win64.  I thought it'd be good to get rid of unneeded complications. This is the way this code used to be a year ago.

File: src/doc/trpl/no-stdlib.md (hunk 3 above)
Comment (alexcrichton) [inline src/doc/trpl/no-stdlib.md]: Can you hide this by default? (e.g. put a `#` in front)

File: src/librustc_back/target/x86_64_pc_windows_gnu.rs (hunk 4 above)
Comment (alexcrichton) [inline src/librustc_back/target/x86_64_pc_windows_gnu.rs]: Ok, sounds good to me

File: src/libstd/rt/dwarf/eh.rs (hunk 5 above)
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Yeah as many links as possible would be nice, even if they're just links to code in other repos. Surely this implementation came from somewhere, and it'd be good to have a trail saying how so!

File: src/libstd/rt/unwind/gcc.rs (hunk 8 above)
Comment (alexcrichton) [inline src/libstd/rt/unwind/gcc.rs]: Ah well, if it passes all the bots then it's fine by me!

File: src/libstd/rt/unwind/gcc.rs (hunk 7 above)
Comment (alexcrichton) [inline src/libstd/rt/unwind/gcc.rs]: Right, thanks!

File: src/libstd/rt/unwind/seh64_gnu.rs (hunk 12 above)
Comment (vadimcn) [inline src/libstd/rt/unwind/seh64_gnu.rs]: > Is this just the same as rust_eh_personality_catch? These two functions look the same except for the expectation of the landing pad existing...

Yes, that the only difference at the moment.  There used to be more when I tried to support running destructors on foreign exceptions.

Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Interesting... I guess this shows how little I know about the ABI of these exceptions on windows! I guess whether a landing pad is a "catch" landing pad is encoded in the landing pad itself and not the personality function?

Overall it basically may mean that we can merge these two functions and only have one?

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+// Define our exception codes:
+// according to http://msdn.microsoft.com/en-us/library/het71c37(v=VS.80).aspx,
+//    [31:30] = 3 (error), 2 (warning), 1 (info), 0 (success)
+//    [29]    = 1 (user-defined)
+//    [28]    = 0 (reserved)
+// we define bits:
+//    [24:27] = type
+//    [0:23]  = magic
+const ETYPE: DWORD = 0b1110_u32 << 28;
+const MAGIC: DWORD = 0x525354; // "RST"
```
Comment (liigo) [inline src/libstd/rt/unwind/seh64_gnu.rs]: `RST` also refers to `reStructuredText`. How about we use `.rs` here?
```
const MAGIC: DWORD = 0x2e7273; // ".rs"
```

File: src/libstd/rt/unwind/gcc.rs (hunk 7 above)
Comment (vadimcn) [inline src/libstd/rt/unwind/gcc.rs]: I'll try to make the equivalent simplification in your `trans_gnu_try`.

**Comment by brson** [comment]
Epic patch.

**Comment by vadimcn** [comment]
Addressed comments and rebased on top of master. r?

File: src/libstd/rt/unwind/mod.rs [hunk 15]
```diff
@@ -76,9 +76,20 @@ use sys_common::mutex::Mutex;
...
+#[path = "seh.rs"] #[doc(hidden)]
+pub mod imp;
+
+// x86_64-pc-windows-gnu
+#[cfg(all(windows, target_arch="x86_64", target_env="gnu"))]
+#[path = "seh64_gnu.rs"] #[doc(hidden)]
 pub mod imp;
-#[cfg(not(target_env = "msvc"))] #[path = "gcc.rs"] #[doc(hidden)]
+
+// x86-pc-windows-gnu and all others
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/mod.rs]: s/x86/i686/

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    }
+    ExceptionContinueSearch
+}
+
+// The `resume` instruction, found at the end of the landing pads, and whose job
+// is to resume stack unwinding, is typically lowered by LLVM into a call to
+// `_Unwind_Resume` routine.  To avoid confusion with the same symbol exported
+// from libgcc, we redirect it to `rust_eh_unwind_resume`.
+// Since resolution of this symbol is done by the linker, `rust_eh_unwind_resume`
+// must be marked `pub` + `#[no_mangle]`.  (Can we make it a lang item?)
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Unfortunately I don't think this can be a lang item because libraries like libcore/libcollections need to reference this (they're generating landing pads) so it needs to have a well known symbol name (e.g. it's used before it's defined).

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+    let er = &*exceptionRecord;
+    let dc = &*dispatcherContext;
+    rtdebug!("rust_eh_personality_catch: code={:X}, flags={:X}, frame={:X}, ip={:X}",
+        er.ExceptionCode, er.ExceptionFlags,
+        establisherFrame as usize, dc.ControlPc as usize);
+
+    if er.ExceptionFlags & EXCEPTION_UNWIND == 0 { // we are in the dispatch phase
+        if er.ExceptionCode == RUST_PANIC {
+            if let Some(lpad) = find_landing_pad(dc) {
+                RtlUnwindEx(establisherFrame,
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: So if I'm understanding this correctly, this call to `RtlUnwindEx` magically jumps back into the function and executes some code? The [documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms680615%28v=vs.85%29.aspx) is a bit scant, and I thought when poking through the libgcc sources that this was basically re-raising the exception, but I also have no idea what's going on here :)

Also, could the `eh_personality` and `eh_personality_catch` functions be merged? Or could the catch one just call the normal one?

Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: (as in the extra assertion in the catch that a landing pad is there doesn't seem worth the duplication)

File: src/rustllvm/RustWrapper.cpp [hunk 18]
```diff
@@ -958,3 +958,27 @@ LLVMRustBuildLandingPad(LLVMBuilderRef Builder,
...
+    CI->setCallingConv(CallingConv::C);
+    // We never expect rust_eh_unwind_resume to return.
+    new UnreachableInst(Ctx, BB);
+
+    // Until DwarfEHPrepare pass has run, _Unwind_Resume is not referenced by any live code
+    // and is subject to dead code elimination.  Here we add _Unwind_Resume to @llvm.globals to prevent that.
+    PointerType *elem_type = PointerType::get(IntegerType::get(Ctx, 8), 0);
+    ArrayType *used_type = ArrayType::get(elem_type, 1);
+    Constant* init = ConstantArray::get(used_type, ArrayRef<Constant*>(ConstantExpr::getBitCast(OrigRewind, elem_type)));
+    GlobalVariable* used = new GlobalVariable(*M, used_type, false, GlobalValue::AppendingLinkage, init, "llvm.used");
```
Comment (alexcrichton) [inline src/rustllvm/RustWrapper.cpp]: Could you be sure to wrap these lines to 80 chars as well?

File: src/rustllvm/RustWrapper.cpp [hunk 19]
```diff
@@ -958,3 +958,27 @@ LLVMRustBuildLandingPad(LLVMBuilderRef Builder,
...
+    used->setSection("llvm.metadata");
+}
```
Comment (alexcrichton) [inline src/rustllvm/RustWrapper.cpp]: Can this all be written in Rust? It's generally quite difficult to keep C++ up to date across LLVM versions and the C API is much more stable.

File: src/libstd/rt/dwarf/mod.rs [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+    // ULEB128 and SLEB128 encodings are defined in Section 7.6 - "Variable
+    // Length Data".
+    pub unsafe fn read_uleb128(&mut self) -> u64 {
+        let mut shift : usize = 0;
+        let mut result : u64 = 0;
+        let mut byte : u8;
+        loop {
+            byte = *self.ptr;
+            self.ptr = self.ptr.offset(1);
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/mod.rs]: Perhaps this could be `let byte = self.read::<u8>()` to consolidate the shifting of `ptr`?

File: src/libstd/rt/dwarf/mod.rs [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Utilities for parsing DWARF-encoded data streams.
+//! See http://www.dwarfstd.org,
+//! DWARF-4 standard, Section 7 - "Data Representation"
+
+#![allow(unused)]
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/mod.rs]: Could you add a small comment to this indicating that this module isn't used on all platforms?

File: src/libstd/rt/dwarf/eh.rs [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+#[inline]
+fn round_up(unrounded: usize, align: usize) -> usize {
+    assert!(align.is_power_of_two());
+    (unrounded + align - 1) & !(align - 1)
+}
+
+unsafe fn read_encoded_pointer(reader: &mut DwarfReader,
+                               context: EHContext,
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Perhaps this could be passed by reference? It looks like this doesn't necessarily need to take ownershipo

File: src/libstd/rt/dwarf/eh.rs [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+        // Callsite table is sorted by cs_start, so if we've passed the ip, we
+        // may stop searching.
+        if ip < func_start + cs_start {
+            break
+        }
+        if ip < func_start + cs_start + cs_len {
+            if cs_lpad != 0 {
+                return Some(lpad_base + cs_lpad);
+            } else {
+                return None;
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Should this and the "not found" case be differentiated? http://www.airs.com/blog/archives/464 seems to indicate that C++ aborts in the "not found" case but simply continues in the case of "`cs_lpad` is 0". May be worth having the extra assertion on the other end?

File: src/libstd/rt/dwarf/eh.rs [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+    let call_site_encoding = reader.read::<u8>();
+    let call_site_table_length = reader.read_uleb128();
+    let action_table = reader.ptr.offset(call_site_table_length as isize);
+    let ip = context.ip-1;
+
+    while reader.ptr < action_table {
+        let cs_start = read_encoded_pointer(&mut reader, context, call_site_encoding);
+        let cs_len = read_encoded_pointer(&mut reader, context, call_site_encoding);
+        let cs_lpad = read_encoded_pointer(&mut reader, context, call_site_encoding);
+        let cs_action = reader.read_uleb128();
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Should we have an assertion that this is 0 most of the time? It sounds like the only nonzero entry should be the catch block, right?

File: src/libstd/rt/dwarf/eh.rs [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+        DW_EH_PE_udata4 => reader.read::<u32>() as usize,
+        DW_EH_PE_udata8 => reader.read::<u64>() as usize,
+        DW_EH_PE_sleb128 => reader.read_sleb128() as usize,
+        DW_EH_PE_sdata2 => reader.read::<i16>() as usize,
+        DW_EH_PE_sdata4 => reader.read::<i32>() as usize,
+        DW_EH_PE_sdata8 => reader.read::<i64>() as usize,
+        _ => panic!()
+    };
+
+    result += match encoding & 0x70 {
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Should this mask be `0xF0`? It looks like this misses the `DW_EH_PE_indirect` case (although it may not actually come up in practice)

File: src/libstd/rt/dwarf/eh.rs [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+        DW_EH_PE_absptr => 0,
+        // relative to address of the encoded value, despite the name
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Well if that's not confusing I don't know what is!

File: src/libstd/rt/dwarf/eh.rs (hunk 25 above)
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Oh I see you handle this case below. Perhaps it could be added to this `match` though?

File: src/libstd/rt/unwind/seh64_gnu.rs [hunk 27]
```diff
@@ -0,0 +1,248 @@
...
+                   &params as *const ULONG_PTR);
+    rtabort!("could not resume unwind");
+}
+
+unsafe fn find_landing_pad(dc: &DISPATCHER_CONTEXT) -> Option<usize> {
+    let eh_ctx = eh::EHContext {
+        ip: dc.ControlPc as usize,
+        func_start: dc.ImageBase as usize + (*dc.FunctionEntry).BeginAddress as usize,
+        text_start: dc.ImageBase as usize,
+        data_start: 0
```
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Perhaps this field could be omitted for now? The `datarel` encoding may want to just panic saying "unimplemented".

File: src/libstd/rt/dwarf/eh.rs [hunk 28]
```diff
@@ -0,0 +1,150 @@
...
+    let ttype_encoding = reader.read::<u8>();
+    if ttype_encoding != DW_EH_PE_omit {
+        // Rust doesn't analyze exception types, so we don't care about the type table
+        reader.read_uleb128();
+    }
+
+    let call_site_encoding = reader.read::<u8>();
+    let call_site_table_length = reader.read_uleb128();
+    let action_table = reader.ptr.offset(call_site_table_length as isize);
+    let ip = context.ip-1;
```
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Could you add a comment to why this `-1` is done here?

**Comment by alexcrichton** [comment]
After going through all the nitty gritty I'm even more impressed now, very nice work @vadimcn!

File: src/libstd/rt/unwind/seh64_gnu.rs (hunk 17 above)
Comment (vadimcn) [inline src/libstd/rt/unwind/seh64_gnu.rs]: > So if I'm understanding this correctly, this call to RtlUnwindEx magically jumps back into the function and executes some code?

Yes, pretty much.  It unwinds stack down to that function's frame, then resumes execution at the specified IP.   The way I chose to implement this is a bit different than what gcc does; their way ended up too complicated, IMO, because they tried to re-use the same personality routine used for DWARF-style unwinding.

> Also, could the eh_personality and eh_personality_catch functions be merged?

I kinda prefer to keep them separate,- for symmetry with other unwinding implementations.  But OK, I'll merge them.

File: src/rustllvm/RustWrapper.cpp (hunk 19 above)
Comment (vadimcn) [inline src/rustllvm/RustWrapper.cpp]: Hm, ok.  Never thought about it from this angle.

File: src/libstd/rt/dwarf/eh.rs (hunk 23 above)
Comment (vadimcn) [inline src/libstd/rt/dwarf/eh.rs]: Right.   But the [libgcc's implementation](https://github.com/gcc-mirror/gcc/blob/master/libgcc/unwind-c.c#L216) actually treats them the same.  Not sure which one is correct.  Not emitting info for an IP range, if no action needs to be taken, could be a valid optimization to reduce LSDA size, IMO.

File: src/libstd/rt/dwarf/eh.rs (hunk 25 above)
Comment (vadimcn) [inline src/libstd/rt/dwarf/eh.rs]: DW_EH_PE_indirect may be combined with any of the above flags, so it's less code this way.

Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Ah ok that makes sense

File: src/libstd/rt/dwarf/eh.rs (hunk 23 above)
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: Hm ok, could you add a comment to that effect?

File: src/libstd/rt/unwind/seh64_gnu.rs (hunk 17 above)
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Oh I think it's fine to keep two personality functions, I just figured it'd be good to reduce the code duplication.

**Comment by vadimcn** [comment]
Addressed review comments.

Also, now that I'm using `llvm.used` to prevent elimination of `_Unwind_Resume` hook by the optimizer, I moved hook injection earlier in trans, and was able to make `rust_eh_unwind_resume` a lang item!

File: src/librustc_trans/trans/common.rs [hunk 29]
```diff
@@ -530,6 +530,51 @@ impl<'a, 'tcx> FunctionContext<'a, 'tcx> {
             }
         }
     }
+
+    pub fn inject_unwind_resume_hook(&self) {
```
Comment (alexcrichton) [inline src/librustc_trans/trans/common.rs]: Could you add a doc comment here explaining what this function is doing?

File: src/librustc_trans/trans/intrinsic.rs [hunk 30]
```diff
@@ -1188,108 +1176,61 @@ fn trans_gnu_try<'blk, 'tcx>(bcx: Block<'blk, 'tcx>,
...
-                                llfn: ValueRef,
-                                lpad_ty: Type,
-                                personality: ValueRef,
-                                output: ty::FnOutput<'tcx>,
-                                dloc: DebugLoc,
-                                invoke: &mut FnMut(Block, Block, Block) -> ValueRef) {
         let (fcx, block_arena);
         block_arena = TypedArena::new();
-        fcx = new_fn_ctxt(ccx, llfn, ast::DUMMY_NODE_ID, false,
+        fcx = new_fn_ctxt(ccx, rust_try, ast::DUMMY_NODE_ID, false,
```
Comment (alexcrichton) [inline src/librustc_trans/trans/intrinsic.rs]: If you're feeling ambitious it looks like the `_Unwind_Resume` shim you created above is much nicer to use than creating all these contexts here, so it may be easier to do that here as well. That being said it's just cleanup so feel free to pass!

File: src/libstd/rt/dwarf/eh.rs (hunk 24 above)
Comment (alexcrichton) [inline src/libstd/rt/dwarf/eh.rs]: (ping on this)

File: src/libstd/rt/unwind/seh64_gnu.rs (hunk 17 above)
Comment (alexcrichton) [inline src/libstd/rt/unwind/seh64_gnu.rs]: (ping on this)

Comment (vadimcn) [inline src/libstd/rt/unwind/seh64_gnu.rs]: Okay, if you insist...

File: src/libstd/rt/dwarf/eh.rs (hunk 24 above)
Comment (vadimcn) [inline src/libstd/rt/dwarf/eh.rs]: ok

Comment (vadimcn) [inline src/libstd/rt/dwarf/eh.rs]: Actually, this is not true - the action for `rust_try`'s landing pad is not 0, though it does point to a null entry in the types table.   I think I'd rather not implement exception type lookup logic just to assert that it's a "catch-all".

**Comment by vadimcn** [comment]
looks ok now?

**Comment by alexcrichton** [comment]
Looks good to me! r=me with a rebase

**Comment by vadimcn** [comment]
rebased. r?

**Comment by vadimcn** [comment]
Broken by last rebase :(

**Comment by vadimcn** [comment]
Works now.  r?

**Comment by Gankro** [comment]
Seems legit:

```

../src/libstd/rt/unwind/gcc.rs:231:38: 231:43 error: use of moved
value: `state` [E0382]
../src/libstd/rt/unwind/gcc.rs:231
__gcc_personality_v0(state, ue_header, context)
                                                                        ^~~~~
../src/libstd/rt/unwind/gcc.rs:225:13: 225:18 note: `state` moved here
because it has type `rt::libunwind::_Unwind_State`, which is
non-copyable
../src/libstd/rt/unwind/gcc.rs:225         if (state as c_int &
uw::_US_ACTION_MASK as c_int)
                                               ^~~~~
error: aborting due to previous error
make: *** [x86_64-unknown-linux-gnu/stage2/lib/rustlib/arm-linux-androideabi/lib/stamp.std]
Error 101
make: *** Waiting for unfinished jobs....

```

**Comment by vadimcn** [comment]
Fixed.  Let's try again?

**Comment by alexcrichton** [comment]
:confetti_ball:
